src/main/java/uk/gov/hmcts/reform/idam/web/AppController.java

```diff
@@ -382,6 +407,92 @@ private String authoriseUser(List<String> cookies, HttpServletRequest httpReques
+     * @should return login view when authorize fails
+     */
+    @PostMapping("/verification")
+    public ModelAndView verification(@ModelAttribute("authorizeCommand") AuthorizeRequest request,
+                               BindingResult bindingResult,
+                               Model model,
+                               HttpServletRequest httpRequest,
+                               HttpServletResponse response) {
+
+        final String ipAddress = ObjectUtils.defaultIfNull(httpRequest.getHeader(X_FORWARDED_FOR), httpRequest.getRemoteAddr());
+
+        final List<String> cookies = Arrays.stream(ofNullable(httpRequest.getCookies()).orElse(new Cookie[] {}))
+            .map(c -> String.format("%s=%s", c.getName(), c.getValue())) // map to: "Idam.AuthId=xyz"
+            .collect(Collectors.toList());
+
+        model.addAttribute(RESPONSE_TYPE, request.getResponse_type());
+        model.addAttribute(STATE, request.getState());
+        model.addAttribute(CLIENT_ID, request.getClient_id());
+        model.addAttribute(REDIRECT_URI, request.getRedirect_uri());
+        model.addAttribute(SCOPE, request.getScope());
+        model.addAttribute(SELF_REGISTRATION_ENABLED, request.isSelfRegistrationEnabled());
+
+        try {
+            final List<String> responseCookies = spiService.submitOtpeAuthentication(cookies, ipAddress, request.getCode());
+            log.info("/verification: Successful OTP submission request - {}", obfuscateEmailAddress(request.getUsername()));
+
+            final String responseUrl = authoriseUser(responseCookies, httpRequest);
+            final boolean loginSuccess = responseUrl != null && !responseUrl.contains("error");
+            if (loginSuccess) {
+                log.info("/verification: Successful login - {}", obfuscateEmailAddress(request.getUsername()));
+                List<String> secureCookies = makeCookiesSecure(responseCookies);
+                secureCookies.forEach(cookie -> response.addHeader(HttpHeaders.SET_COOKIE, cookie));
+                return new ModelAndView("redirect:" + responseUrl);
+            } else {
+                log.info("/verification: There is a problem while logging in user - {}", obfuscateEmailAddress(request.getUsername()));
+                model.addAttribute(HAS_LOGIN_FAILED, true);
+                bindingResult.reject("Login failure");
+                return new ModelAndView("redirect:/login", model.asMap());
+            }
+        } catch (HttpClientErrorException | HttpServerErrorException he) {
+            log.info("/verification: Login failed for user - {}", obfuscateEmailAddress(request.getUsername()));
+            if (HttpStatus.UNAUTHORIZED == he.getStatusCode()) {
+
+                ErrorResponse error = new ErrorResponse();
+                if (he.getResponseBodyAsString() != null) {
+                    try {
+                        error = objectMapper.readValue(he.getResponseBodyAsString(), ErrorResponse.class);
```

Comment:
Might be better to have a helper method that fails silently for this. Otherwise we're gonna end up with lots of uncovered conditions in unit tests.